ARCs/arc-0024.md

```diff
@@ -0,0 +1,485 @@
+description, and that the ARC-4 Contract section be updated to
+include locals and globals:
+
+```typescript
+interface Contract {
+  /** A user-friendly name for the contract */
+  name: string;
+  /** Optional, user-friendly description for the interface */
+  desc?: string;
+  /**
+   * Optional object listing the contract instances across different networks
+   */
+  networks?: {
+    /**
+     * The key is the base64 genesis hash of the network, and the value contains
+     * information about the deployed contract in the network indicated by the
+     * key
+     */
+    [network: string]: {
+      /** The app ID of the deployed contract in this network */
+      appID: number;
+    }
+  }
+  /** All of the methods that the contract implements */
+  methods: Method[];
+
+  /** All local variables intended for consumption as part of this contract **/
+  locals: Variable[];
+
+  /** All global variables intended for consumption as part of this contract **/
+  globals: Variable[];
+
+}
+```
+
+There needs to be a _Variable_ specification too, following similar style as
+the _Method_ entity:
+
+```typescript
+interface Variable {
+  /** The name of the variable **/
+  name: string;
+  /** Optional, user-friendly description for the variable **/
+  desc?: string;
+  /** The type of the variable **/
+  type: string;
+}
+```
+
+For example:
+
+```json
+{
+  "name": "SugarSupplierContract",
+  "desc": "Sugar-Decentralised Inc.'s epic sugar contract. All our sugar is produced by people who only eat sugar! Sustainability is our middle name.",
+  "networks": {
+    "wGHE2Pwdvd7S12BL5FaOP20EGYesN73ktiC1qzkkit8=": { "appID": 1234 },
+    "SGO1GKSzyE7IEPItTxCByw9x8FmnrCDexi9/cOUJOiI=": { "appID": 5678 }
+  },
+  "methods": [
+    {
+      "name": "ShippingCost",
+      "desc": "Get a sugar shipping quote here ",
+      "args": [
+        { "type": "uint64", "name": "a", "desc": "Volume in kilos" }
+      ],
+      "returns": { "type": "uint64", "desc": "Your shipping price in Algo" }
+    }
+  ],
+  "locals": [ 
+      {
+      "name": "YourPrice",
+      "desc": "Call the ShippingCost contract with proposed annual volume in the Note field to get a wholesaler price. ",
+      "type": "uint64"
+    }  
+  ],
+   "globals": [ 
+      {
+      "name": "SugarPrice",
+      "desc": "Our public sugar retail price in Algo ",
+      "type": "uint64"
+    }  
+  ]
+}
+```
+
+### **2. Human and machine readability**
+
+JSON formats, as above, are both human and machine readable.
+
+The specification MUST always be both human and machine readable.
+
+Extensions to this specification MUST NOT ever include machine-only or binary formats.
+
+### **3. Extensibility**
+
+It is difficult to know what the Application Specification will be expected
+to contain in the future. One thing that is reasonably certain though
+is that everyone would benefit if the information about the Application
+was not scattered across different places in different formats.
+
```

Comment:
I think this can be removed if, instead, we extend `Contract` in the same way, in future ARCs.  A future ARC-88 would add another optional key "ARC88" to the `Contract` schema. Old code would ignore it, new code would use it.

But we avoid the complexity of this new envelope which is currently empty.  JSON is a pretty goo envelope for unknown future data already.